sql/api/src/main/scala/org/apache/spark/sql/types/UpCastRule.scala

```diff
@@ -41,10 +41,9 @@ private[sql] object UpCastRule {
     case (TimestampNTZType, TimestampType) => true
     case (TimestampType, TimestampNTZType) => true
 
-    case (_: AtomicType, CharType(_) | VarcharType(_)) => false
-    case (_: CalendarIntervalType, CharType(_) | VarcharType(_)) => false
-    case (_: AtomicType, _: StringType) => true
-    case (_: CalendarIntervalType, _: StringType) => true
+    case (s1: StringType, s2: StringType) => StringHelper.isMoreConstrained(s1, s2)
```

Comment:
shall we allow cast varchar(10) to varchar(5) since it's possible that all the actual string values do not exceed length 5?

Reply:
nvm, this is upcast